lib/teiserver_web/controllers/o_auth/authorize_controller.ex

```diff
@@ -0,0 +1,104 @@
+defmodule TeiserverWeb.OAuth.AuthorizeController do
+  use TeiserverWeb, :controller
+
+  alias Teiserver.OAuth
+
+  def authorize(conn, params) when not is_map_key(params, "client_id") do
+    bad_request(conn, "missing client_id")
+  end
+
+  def authorize(conn, %{"client_id" => client_id} = params) do
+    case OAuth.get_application_by_uid(client_id) do
+      nil ->
+        bad_request(conn, "invalid client_id")
+
+      app ->
+        conn
+        |> render("authorize.html",
+          app_name: app.name,
+          params: params
+        )
+    end
+  end
+
+  def authorize(conn, _params) do
+    conn |> render("bad_request.html")
+  end
+
+  def generate_code(conn, params) when not is_map_key(params, "client_id") do
+    bad_request(conn, "missing client_id")
+  end
+
+  def generate_code(conn, params) when not is_map_key(params, "redirect_uri") do
+    bad_request(conn, "missing redirect_uri")
+  end
+
+  def generate_code(conn, %{"client_id" => client_id, "redirect_uri" => redirect_uri} = params) do
+    case OAuth.get_application_by_uid(client_id) do
+      nil ->
+        bad_request(conn, "invalid client_id")
+
+      app ->
+        case OAuth.get_redirect_uri(app, redirect_uri) do
+          {:error, _} -> bad_request(conn, "invalid redirection url")
+          {:ok, redir_url} -> do_generate_code(conn, app, redir_url, params)
+        end
+    end
+  end
+
+  defp do_generate_code(conn, app, redir_url, params) do
+    cond do
+      Map.get(params, "response_type") != "code" ->
+        error_redirect(conn, redir_url, "unsupported_response_type", "only code is supported")
+
+      Map.get(params, "code_challenge_method") != "S256" ->
+        error_redirect(conn, redir_url, "invalid_request", "only S256 is supported")
+
+      Map.get(params, "code_challenge") == nil ->
+        error_redirect(conn, redir_url, "invalid_request", "a code challenge must be provided")
+
+      true ->
+        code_params = %{
+          id: app.id,
+          redirect_uri: URI.to_string(redir_url),
+          scopes: app.scopes,
+          challenge: Map.get(params, "code_challenge"),
+          challenge_method: "S256"
+        }
+
+        case OAuth.create_code(conn.assigns.current_user, code_params) do
+          {:error, _} ->
+            error_redirect(conn, redir_url, "server_error", "something went wrong")
```

Comment:
`state` is also required in error response if it was present in the request.

Reply:
I've changed the [error_redirect function](https://github.com/geekingfrog/teiserver/blob/3514c247d065f9b028c50923e390ecc919d9513f/lib/teiserver_web/controllers/o_auth/authorize_controller.ex#L110-L130) to make it so.